src/api/server.rkt

```diff
@@ -53,75 +82,394 @@
+       [(list 'improve handler)
+        (define improved-list
+          (for/list ([(job-id result) completed-work]
+                     #:when (equal? (hash-ref result 'command) "improve"))
+            (get-table-data-from-hash result (make-path job-id))))
+        (place-channel-put handler improved-list)]
+       ; Start a worker on a job. Unless the job-id is marked as finished then move to next state.
+       [(list 'start self command job-id)
+        (if (hash-has-key? completed-work job-id)
+            (place-channel-put self (list 'finished job-id (hash-ref completed-work job-id)))
+            (let ([worker (make-worker)])
+              ; Maybe this should be worker-id and we should pre allocate workers based on threads available.
+              (hash-set! workers job-id worker)
+              (when verbose
+                (eprintf "Starting worker [~a] on [~a].\n"
+                         job-id
+                         (test-name (herbie-command-test command))))
+              (place-channel-put worker (list 'apply self command job-id))))]
+       ; Job is finished notified the waiting list with the result of the job.
+       [(list 'finished job-id result)
+        (when verbose
+          (eprintf "Job ~a finished, saving result.\n" job-id))
+        ; Notifed job has been completed, save the result.
+        ; let GC collect worker 🤞.
+        (hash-set! completed-work job-id result)
+        (hash-remove! workers job-id)
+        (define maybe-wait-list (hash-ref waiting job-id #f))
+        (when maybe-wait-list
+          (when verbose
+            (eprintf "waiting job ~a completed\n" job-id))
+          (for ([waiting maybe-wait-list])
+            (place-channel-put waiting result))
+          (hash-remove! waiting job-id))]
+       ; Check if work is completed, returns the result or #f
+       [(list 'check job-id handler) (place-channel-put handler (hash-ref completed-work job-id #f))]
+       ; Pass a place-channel `handler` to the receptionist to be notified on when a job is complete.
+       [(list 'wait job-id handler)
+        ; first we add the handler to the wait list.
+        (if (false? (hash-ref waiting job-id #f))
+            (hash-set! waiting job-id (list handler))
+            (let ([wait-list (hash-ref waiting job-id)])
+              (hash-set! waiting job-id (append wait-list (list handler)))))
+        (define result (hash-ref completed-work job-id #f))
+        (when verbose
+          (eprintf "Waiting for job: ~a\n" job-id))
+        ; check if the job is completed oor not.
+        (unless (false? result)
+          ; we have a result to send.
+          (let ([maybe-wait-list (hash-ref waiting job-id #f)])
+            (unless (false? maybe-wait-list)
+              ; we have a waiting list to notify
+              (for ([waiting maybe-wait-list])
+                (place-channel-put waiting result))
+              (hash-remove! waiting job-id))))]))))
```

Comment:
I think this is fine but let's think this through—if the job is complete, then the wait list must have been emptied when the job finished. The only other msg that can add to the wait list is `'wait` and it also empties the wait list. So the wait list is always empty if the job is complete. So I think the `'wait` logic can be must easier:

- If the job is completed, respond immediately
- If not, add it to the wait list

Then only the job actually finishing would clear the wait list.

Reply:
Hmm, I will have to come back to this. Having trouble getting this to work with two of the same jobs coming in like `testApi.mjs` does with `/sample`. I might also need to fix `'start` 🤔